File: java/src/main/java/com/jpmanjarres/hackerrank/algorithms/sorting/Quicksort2Sorting.java
/**
 * Evoki.co - Plume Works Copyright (c) 2013 - 2015 by Jean Paul Manjarres Correal - All Rights
 * Reserved. Propietary and Confidential - This source code is not for redistribution
 * http://evoki.co Date: 26/11/2015
 *
 */
package com.jpmanjarres.hackerrank.algorithms.sorting;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/**
 * @author <a href="mailto:paul.manjarres@gmail.com">Jean Paul Manjarres Correal. </a> 26/11/2015
 *
 */
public class Quicksort2Sorting {

	static void quickSort(int[] ar) {


		int partition = ar[0];

		List<Integer> left = new ArrayList<Integer>();
		List<Integer> right = new ArrayList<Integer>();

		for (int i = 1; i < ar.length; i++) {

			if(ar[i]< partition){
				left.add(ar[i]);
			}else{
				right.add(ar[i]);
			}

		}






	}

	static void printArray(int[] ar) {
		for (int n : ar) {
			System.out.print(n + " ");
		}
		System.out.println("");
	}

	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		int n = in.nextInt();
		int[] ar = new int[n];
		for (int i = 0; i < n; i++) {
			ar[i] = in.nextInt();
		}
		quickSort(ar);
	}

}

File: coding-exercises-kotlin/src/main/kotlin/org/yagamipaul/hackerrank/algorithms/sorting/Quicksort2Sorting.kt
package org.yagamipaul.hackerrank.algorithms.sorting

import java.util.*
import kotlin.collections.ArrayList


/**
 * Quicksort 2 - Sorting
 * https://www.hackerrank.com/challenges/quicksort2/problem
 * Jean Paul Manjarres Correal <paul.manjarres@gmail.com>
 */
fun main(args: Array<String>) {

    val sc = Scanner(System.`in`)
    val n = sc.nextInt()
    val arr = IntArray(n)

    for( i in 0 until n){
        arr[i] = sc.nextInt()
    }
    solve(arr)

}

/**
 * Solves the problem in a recursive way
 */
private fun solve(arr:IntArray): IntArray{

    // If the array is only 1 element or empty, is already sorted.
    if(arr.size == 1 || arr.isEmpty()){
        return arr
    }

    val pivot = arr[0]

    val leftArr = ArrayList<Int>()
    val rightArr = ArrayList<Int>()


    for(i in 1 until arr.size){
        when{
            arr[i] < pivot -> leftArr.add(arr[i])
            arr[i] > pivot -> rightArr.add(arr[i])
        }
    }

   // println("left: ${printArray(leftArr.toIntArray())}")
   // println("Right: ${printArray(rightArr.toIntArray())}")

    val merged = intArrayOf(*solve(leftArr.toIntArray()),pivot,*solve(rightArr.toIntArray()))
    println("${printArray(merged)}")
    return merged

}


private fun printArray(arr: IntArray) : String{
    var str = ""
    for(i in arr){
        str+="$i "
    }
    return str
}